appengine/standard/migration/incoming/main.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Authenticate requests coming from other App Engine instances.
+"""
+
+# [START gae_python_app_identity_incoming]
+from google.oauth2 import id_token
+from google.auth.transport import requests
+
+import webapp2
+
+
+def get_app_id(request):
+    # Requests from App Engine Standard for Python 2.7 will include a
+    # trustworthy X-Appengine-Inbound-Appid. Other requests won't have
+    # that header, as the App Engine runtime will strip it out
+    incoming_app_id = request.headers.get(
+        'X-Appengine-Inbound-Appid', None)
+    if incoming_app_id is not None:
+        return incoming_app_id
+
+    # Other App Engine apps can get an ID token for the App Engine default
+    # service account, which will identify the application ID. They will
+    # have to include at token in an Authorization header to be recognized
+    # by this method.
+    auth_header = request.headers.get('Authorization', None)
+    if auth_header is None:
+        return None
+
+    # The auth_header must be in the form Authorization: Bearer token.
+    bearer, token = auth_header.split()
+    if bearer.lower() != 'bearer':
+        return None
+
+    try:
+        info = id_token.verify_oauth2_token(token, requests.Request())
+        service_account_email = info['email']
+        incoming_app_id, domain = service_account_email.split('@')
+        if domain != 'appspot.gserviceaccount.com':  # Not App Engine svc acct
+            return None
+        else:
+            return incoming_app_id
```

Comment:
May also want to verify the audience & issuer here. 

> The (Python) verify_oauth2_token function verifies the JWT signature, the aud claim, and the exp claim. You must also verify the iss claim and the hd claim (if applicable) by examining the object that verify_oauth2_token returns. If multiple clients access the backend server, also manually verify the aud claim. ([ref](https://developers.google.com/identity/sign-in/web/backend-auth?_ga=2.135586742.644768118.1588289335-2109548938.1587593579#using-a-google-api-client-library))

Reply:
I thought long and hard about verifying the audience, but eventually decided that I was showing how to migrate from the old X-Appengine-Inbound-Appid approach, which did not perform that check.

The Google verify_oauth2_token verifies tokens issued by Google's OAuth2 server, so I don't think there needs to a separate check of the issuer.

Reply:
SGTM. 

The text & samples on their [website](https://developers.devsite.corp.google.com/identity/sign-in/web/backend-auth#using-a-google-api-client-library) for other languages vs. Python are a little misleading. I will file a bug.

Reply:
> It's possible that those other language libraries don't do the same issuer
verification automatically.

@engelke Hold on, I thought the text implied that Python doesn't verify issuer automatically.